File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/ui/flashcard/FlashcardViewModel.kt
package com.jworks.kanjiquest.android.ui.flashcard

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jworks.kanjiquest.core.domain.model.FlashcardDeckGroup
import com.jworks.kanjiquest.core.domain.model.FlashcardEntry
import com.jworks.kanjiquest.core.domain.model.Kanji
import com.jworks.kanjiquest.core.domain.repository.FlashcardRepository
import com.jworks.kanjiquest.core.domain.repository.KanjiRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class FlashcardItem(
    val entry: FlashcardEntry,
    val kanji: Kanji
)

data class FlashcardUiState(
    val items: List<FlashcardItem> = emptyList(),
    val deckGroups: List<FlashcardDeckGroup> = emptyList(),
    val selectedDeckId: Long = 1,
    val isLoading: Boolean = true,
    val showCreateDeckDialog: Boolean = false,
    val editingDeck: FlashcardDeckGroup? = null
)

@HiltViewModel
class FlashcardViewModel @Inject constructor(
    private val flashcardRepository: FlashcardRepository,
    private val kanjiRepository: KanjiRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(FlashcardUiState())
    val uiState: StateFlow<FlashcardUiState> = _uiState.asStateFlow()

    init {
        loadDecks()
    }

    private fun loadDecks() {
        viewModelScope.launch {
            flashcardRepository.ensureDefaultDeck()
            val groups = flashcardRepository.getAllDeckGroups()
            val selectedId = _uiState.value.selectedDeckId
            val effectiveId = if (groups.any { it.id == selectedId }) selectedId else groups.firstOrNull()?.id ?: 1
            _uiState.value = _uiState.value.copy(deckGroups = groups, selectedDeckId = effectiveId)
            loadDeck(effectiveId)
        }
    }

    private fun loadDeck(deckId: Long) {
        viewModelScope.launch {
            val entries = flashcardRepository.getFlashcardsByDeck(deckId)
            val items = entries.mapNotNull { entry ->
                val kanji = kanjiRepository.getKanjiById(entry.kanjiId)
                if (kanji != null) FlashcardItem(entry, kanji) else null
            }
            _uiState.value = _uiState.value.copy(
                items = items,
                selectedDeckId = deckId,
                isLoading = false
            )
        }
    }

    fun selectDeck(deckId: Long) {
        _uiState.value = _uiState.value.copy(isLoading = true)
        loadDeck(deckId)
    }

    fun createDeck(name: String) {
        viewModelScope.launch {
            val newId = flashcardRepository.createDeckGroup(name)
            val groups = flashcardRepository.getAllDeckGroups()
            _uiState.value = _uiState.value.copy(
                deckGroups = groups,
                showCreateDeckDialog = false
            )
            selectDeck(newId)
        }
    }

    fun renameDeck(deckId: Long, newName: String) {
        viewModelScope.launch {
            flashcardRepository.renameDeckGroup(deckId, newName)
            val groups = flashcardRepository.getAllDeckGroups()
            _uiState.value = _uiState.value.copy(
                deckGroups = groups,
                editingDeck = null
            )
        }
    }

    fun deleteDeck(deckId: Long) {
        viewModelScope.launch {
            flashcardRepository.deleteDeckGroup(deckId)
            val groups = flashcardRepository.getAllDeckGroups()
            if (groups.isEmpty()) {
                flashcardRepository.ensureDefaultDeck()
                val refreshedGroups = flashcardRepository.getAllDeckGroups()
                _uiState.value = _uiState.value.copy(deckGroups = refreshedGroups, editingDeck = null)
                selectDeck(refreshedGroups.first().id)
            } else {
                _uiState.value = _uiState.value.copy(deckGroups = groups, editingDeck = null)
                if (deckId == _uiState.value.selectedDeckId) {
                    selectDeck(groups.first().id)
                }
            }
        }
    }

    fun showCreateDeckDialog() {
        _uiState.value = _uiState.value.copy(showCreateDeckDialog = true)
    }

    fun dismissCreateDeckDialog() {
        _uiState.value = _uiState.value.copy(showCreateDeckDialog = false)
    }

    fun showEditDeckDialog(deck: FlashcardDeckGroup) {
        _uiState.value = _uiState.value.copy(editingDeck = deck)
    }

    fun dismissEditDeckDialog() {
        _uiState.value = _uiState.value.copy(editingDeck = null)
    }

    fun removeFromDeck(kanjiId: Int) {
        viewModelScope.launch {
            flashcardRepository.removeFromDeck(_uiState.value.selectedDeckId, kanjiId)
            _uiState.value = _uiState.value.copy(
                items = _uiState.value.items.filter { it.entry.kanjiId != kanjiId }
            )
        }
    }

    fun refresh() {
        loadDecks()
    }
}

File: ios-app/KanjiQuest/Screens/Flashcard/FlashcardViewModel.swift
import Foundation
import SharedCore

struct FlashcardItem: Identifiable {
    var id: Int32 { entry.kanjiId }
    let entry: FlashcardEntry
    let kanji: Kanji?
}

@MainActor
final class FlashcardViewModel: ObservableObject {
    @Published var items: [FlashcardItem] = []
    @Published var deckGroups: [FlashcardDeckGroup] = []
    @Published var selectedDeckId: Int64 = 1
    @Published var isLoading = true
    @Published var showCreateDeckDialog = false
    @Published var editingDeck: FlashcardDeckGroup? = nil

    private var flashcardRepository: FlashcardRepository?
    private var kanjiRepository: KanjiRepository?

    func load(container: AppContainer) {
        flashcardRepository = container.flashcardRepository
        kanjiRepository = container.kanjiRepository
        loadDecks()
    }

    func loadDecks() {
        Task {
            isLoading = true
            try? await flashcardRepository?.ensureDefaultDeck()
            deckGroups = (try? await flashcardRepository?.getAllDeckGroups()) ?? []

            // Select first deck if current not in list
            if !deckGroups.contains(where: { $0.id == selectedDeckId }), let first = deckGroups.first {
                selectedDeckId = first.id
            }

            await loadDeck(selectedDeckId)
            isLoading = false
        }
    }

    func loadDeck(_ deckId: Int64) async {
        let entries = (try? await flashcardRepository?.getFlashcardsByDeck(deckId: deckId)) ?? []
        var newItems: [FlashcardItem] = []
        for entry in entries {
            let kanji = try? await kanjiRepository?.getKanjiById(id: entry.kanjiId)
            newItems.append(FlashcardItem(entry: entry, kanji: kanji))
        }
        items = newItems
    }

    func selectDeck(_ deck: FlashcardDeckGroup) {
        selectedDeckId = deck.id
        isLoading = true
        Task {
            await loadDeck(deck.id)
            isLoading = false
        }
    }

    func createDeck(name: String) {
        Task {
            let _ = try? await flashcardRepository?.createDeckGroup(name: name)
            loadDecks()
            showCreateDeckDialog = false
        }
    }

    func renameDeck(_ deck: FlashcardDeckGroup, to name: String) {
        Task {
            try? await flashcardRepository?.renameDeckGroup(id: deck.id, name: name)
            loadDecks()
            editingDeck = nil
        }
    }

    func deleteDeck(_ deck: FlashcardDeckGroup) {
        Task {
            try? await flashcardRepository?.deleteDeckGroup(id: deck.id)
            if selectedDeckId == deck.id, let first = deckGroups.first(where: { $0.id != deck.id }) {
                selectedDeckId = first.id
            }
            loadDecks()
            editingDeck = nil
        }
    }

    func removeFromDeck(_ item: FlashcardItem) {
        Task {
            try? await flashcardRepository?.removeFromDeck(deckId: selectedDeckId, kanjiId: item.entry.kanjiId)
            await loadDeck(selectedDeckId)
        }
    }

    func refresh() { loadDecks() }
}
